### PR TITLE
issue/8: Save comments to mongodb

### LINKED ISSUE #8: Save comments to MongoDB
**Acceptance criteria**
- Comment service can connect to a MongoDB
- Comment service saved the comment object to the database instead of logging it out
- It works within the docker compose file in flam-flam/helm-charts

### PR DESCRIPTION
**Pull request by fratcliffe**
Closes #8 
Screenshot 2023-04-30 at 17 18 38
Screenshot 2023-04-30 at 17 18 50

### CODE REVIEW

File: src/db.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const mongoose = require("mongoose");
+
+const connectionString =
+  process.env.MONGO_CONNECTION_STRING ||
+  "mongodb://rootuser:examplepasswordikkO2Y@localhost:27017/flam";
+
+// Number of milliseconds to keep trying to establish a db connection
+const serverTimeout = process.env.MONGO_SERVER_TIMEOUT || 30000;
+// Number of milliseconds of inactivity after which to close sockets
+const socketTimeout = process.env.MONGO_SOCKET_TIMEOUT || 45000;
```
Comment (danyloid) [inline src/db.js:10]: I would suggest using connection string to configure everything that you can configure with connection string :) 
Here is the relevant section: https://www.mongodb.com/docs/manual/reference/connection-string/#timeout-options

File: tests/mongo.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  try {
+    await mongoose.connection.close();
+    if (mongod) {
+      await mongod.stop();
+    }
+  } catch (err) {
+    console.log(err);
+    process.exit(1);
+  }
+};
```
Comment (danyloid) [inline tests/mongo.js:47]: Partially mentioned this in the comment on the package file - I would suggest using a separate helper here.

File: tests/mongo.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        .post("/api/comments")
+        .send(dbComment)
+        .expect(200)
+        .expect(function (res) {
+          res.body == dbComment;
+        })
+        .catch((error) => {
+          console.log("MongoDB ERROR: ", error);
+        });
+      done();
```
Comment (danyloid) [inline tests/mongo.js:71]: Right now you probably are seeing an error along the lines of `Cannot log after tests are done. Did you forget to wait for something async in your test?`

The particular syntax seems to be [promise ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) based - meaning that `expect` and `catch` functions would be likely called after `done` is called. 

There are two ways about it - you could declare function as `async` and use `await` to make execution linear - I believe that was used in the example you've referenced (https://medium.com/weekly-webtips/express-js-testing-mocking-mongodb-46c3797a201), look at the `async` & `await` usage

image


Or you could add `done()` call to the `.finally()` block at the end of the chain.

**Review by danyloid — COMMENTED** [review]
Wanted to share few hints, figured commenting with the link to file / line is the easiest way to go.

File: package.json [hunk 4]
```diff
@@ -8,14 +8,19 @@
     "main": "src/server.js",
     "scripts": {
         "start": "node src/server.js",
-        "test": "mocha tests/*.js --exit"
+        "test-orig": "mocha tests/*.js --exit",
+        "test": "jest tests/mongo.js --exit"
```
[the remark this answers is not in the document]
Comment (danyloid) [reply package.json:12]: Updated the comment a bit

**Comment by timberhill** [comment]
I seems to be getting errors when submission service is connecting to mongodb. Looks like I did not configure it correctly.

Mongo container error:
```
Could not connect to database using connectionString: mongodb://rootuser:examplepasswordikkO2Y@mongodb:27017/"
```

.env:
```
REDDIT_CLIENT_ID=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
REDDIT_CLIENT_SECRET=bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
MONGO_INITDB_ROOT_USERNAME=rorootuserot
MONGO_INITDB_ROOT_PASSWORD=examplepasswordikkO2Y
DISPATCHER_BRANCH=main
COMMENT_BRANCH=main
SUBMISSION_BRANCH=main
COMMENT_BRANCH=issue/8
MONGO_CONNECTION_STRING=mongodb://rootuser:examplepasswordikkO2Y@mongodb:27017
```

```bash
~/repo/flam-flam/dev-stack issue/4 ❯ docker ps                                                                                                                                                                timberhill@zenbook
CONTAINER ID   IMAGE                           COMMAND                  CREATED         STATUS                            PORTS                                            NAMES
5f8cf2349d22   dispatcher-service              "python3 -um app"        2 minutes ago   Up 2 minutes                                                                       compose-dispatcher-1
4b4f99028c60   mongo-express                   "tini -- /docker-ent…"   2 minutes ago   Restarting (0) 9 seconds ago                                                       compose-mongo-express-1
1c6412cd965d   comment-service                 "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes                      0.0.0.0:8000->8000/tcp, :::8000->8000/tcp        compose-comment-1
d5312a79878e   submission-service              "dotnet flamflam.Sub…"   2 minutes ago   Up 2 minutes (health: starting)   443/tcp, 0.0.0.0:8001->80/tcp, :::8001->80/tcp   compose-submission-1
03813e655c5a   mongo:6.0                       "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes                      0.0.0.0:27017->27017/tcp, :::27017->27017/tcp    compose-mongodb-1
de887473688f   moby/buildkit:buildx-stable-1   "buildkitd"              7 weeks ago     Up 13 minutes                                                                      buildx_buildkit_cool_nash0
```

**Comment by timberhill** [comment]
It works! Tested with the dev-stack, data does appear in the database 👍 🦆

**Review by timberhill — APPROVED** [review]

**Pull request merged**